Only pass extra create_trail options if set in module params

kwerey opened the pull request:
##### SUMMARY

Fixes #34700  - the Amazon CloudTrail service in China (and I think some other regions) does not support the EnableLogFileValidation and KmsKeyId parameters.

The module currently defaults to setting the former as false and passing the latter as an empty string, which causes an UnsupportedOperationException.

This PR takes out the empty-string defaults, and lets the CloudTrail API default log file validation to false (see [here](https://docs.aws.amazon.com/awscloudtrail/latest/APIReference/API_CreateTrail.html#awscloudtrail-CreateTrail-request-EnableLogFileValidation) for the API reference).

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

cloudtrail

##### ANSIBLE VERSION
```
$ ansible --version
ansible 2.5.0 (cloudtrail-china-support 4913991103) last updated 2018/01/11 17:34:41 (GMT +100)
  config file = /Users/nbailey/.ansible.cfg
  configured module search path = [u'/Users/nbailey/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/nbailey/ansible/lib/ansible
  executable location = /Users/nbailey/ansible/bin/ansible
  python version = 2.7.10 (default, Oct 23 2015, 19:19:21) [GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.59.5)]
```

Fix verified in US and China AWS regions; happy to adjust formatting if there are Ansible module-writing conventions I'm missing.

kwerey committed:
Fixes #34700, only passes extra create_trail options when explicitly set

tedder commented on lib/ansible/modules/cloud/amazon/cloudtrail.py:
```diff
@@ -492,6 +485,9 @@ def main():
         ct_params['CloudWatchLogsLogGroupArn'] = module.params['cloudwatch_logs_log_group_arn']
 
+    if module.params['enable_log_file_validation']:
+        ct_params['EnableLogFileValidation'] = module.params['enable_log_file_validation']
```
Hmm. The problem with this is that it isn't possible to turn off log file validation, making this fail the idempotency test (you can't assert that log file validation is disabled). I'm not sure if the 'bool' param allows for the param to be [omitted](http://docs.ansible.com/ansible/latest/playbooks_filters.html#omitting-parameters) or if it defaults to false; if you can't test that, I'll try to get to it in the next few days/weeks.

kwerey replied:
Huh... I don't think I totally follow, sorry. If you tried to assert it was disabled, anything getting info from 'get-trail-status' would report that it is disabled whether it's set to false or not specified, because "disabled" is the default of the service itself for new trails.

Using default(omit) in a task doesn't remove module defaults, I don't think - just only passes task params if they're set. Happy to test tomorrow UK time.

tedder replied:
Here's the use case:

- enable cloudtrail, set log file validation=true
- set log file validation=false

Given the conditional in 488, validation would still be enabled.

shepdelacreme replied:
I just tested this and can confirm that once you enable log file validation you can't set it back to False due to the conditional on 488.

`AnsibleModule` by default sets parameters to None if they are not specified and do not have a Default set, regardless of their type. So in the current code I think the above behavior can be avoided by changing line 488 to:
```python
if module.params['enable_log_file_validation'] is not None:
    ct_params['EnableLogFileValidation'] = module.params['enable_log_file_validation']
```

kwerey replied:
D'oh - yup, that makes perfect sense by the light of day. Will update + fix the check mode.

tedder replied:
No worries. Thanks for your patience.

shepdelacreme commented:
Module fails in check mode due to a KeyError from the code on line  https://github.com/kwerey/ansible/blob/4913991103e5aeb38e609745ca2535daec705134/lib/ansible/modules/cloud/amazon/cloudtrail.py#L598 

```
Traceback (most recent call last):
  File "/var/folders/sk/k6tqwkbj59x4glzs2hzt2vzrbm012h/T/ansible_XcfNj0/ansible_module_cloudtrail.py", line 617, in <module>
    main()
  File "/var/folders/sk/k6tqwkbj59x4glzs2hzt2vzrbm012h/T/ansible_XcfNj0/ansible_module_cloudtrail.py", line 602, in main
    trail.pop('EnableLogFileValidation')
KeyError: 'EnableLogFileValidation'

fatal: [localhost]: FAILED! => {
    "attempts": 1, 
    "changed": false, 
    "module_stderr": "Traceback (most recent call last):\n  File \"/var/folders/sk/k6tqwkbj59x4glzs2hzt2vzrbm012h/T/ansible_XcfNj0/ansible_module_cloudtrail.py\", line 617, in <module>\n    main()\n  File \"/var/folders/sk/k6tqwkbj59x4glzs2hzt2vzrbm012h/T/ansible_XcfNj0/ansible_module_cloudtrail.py\", line 602, in main\n    trail.pop('EnableLogFileValidation')\nKeyError: 'EnableLogFileValidation'\n", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE", 
    "rc": 0
}
```

kwerey committed:
Fix check mode, fix param validation

kwerey commented:
D'oh, yup, that makes perfect sense by the light of day, @tedder .

Updated this PR so it checks the module parameter is not None instead, and updated the check mode mock return values so it sets ct_params['EnableLogFileValidation'] to false if it wasn't set from the module params.

Does this approach make sense?

I'm happy to do something else otherwise, though I'm suspicious of anything that adds checks like `if region = 'cn-north-1'`: there's now a second China region and there'll probably be other similiarly nonstandard service endpoint behaviours in the future...

tedder commented on lib/ansible/modules/cloud/amazon/cloudtrail.py:
```diff
@@ -595,7 +595,9 @@ def main():
-            trail['LogFileValidationEnabled'] = ct_params['EnableLogFileValidation']
+            if 'EnableLogFileValidation' not in ct_params:
+                ct_params['EnableLogFileValidation'] = False
+            trail['EnableLogFileValidation'] = ct_params['EnableLogFileValidation']
```
This works in China? I thought the param would have to be missing.

kwerey replied:
This isn't making a real request to Cloudtrail in China - it's constructing the spoof return values from a request (hence the fake ARN).

tedder replied:
Aha. I need more coffee. Thanks for doing this.

kwerey replied:
If I'm reading right, what's happening is that the check mode opens an STS connection a few lines up to make sure you've passed real credentials and it really can connect to the region, but then basically makes up a fake API return output right here... 

```
$ ./hacking/test-module -m lib/ansible/modules/cloud/amazon/cloudtrail.py -a "name='test-trail-1' s3_bucket_name='cn-sh-cnn1-cloud-trail' region='cn-north-1' state=present profile='<redacted>'" -c
* including generated source, if any, saving to: /Users/nbailey/.ansible_module_generated
* ansiballz module detected; extracted module source to: /Users/nbailey/debug_dir
***********************************
RAW OUTPUT

{"invocation": {"module_args": {"profile": "<redacted>", "aws_secret_key": null, "aws_access_key": null, "include_global_events": true, "name": "test-trail-1", "security_token": null, "tags": {}, "validate_certs": true, "region": "cn-north-1", "kms_key_id": null, "cloudwatch_logs_log_group_arn": null, "s3_bucket_name": "cn-sh-cnn1-cloud-trail", "is_multi_region_trail": false, "state": "present", "cloudwatch_logs_role_arn": null, "ec2_url": null, "enable_log_file_validation": null, "sns_topic_name": null, "s3_key_prefix": null, "enable_logging": true}}, "trail": {"name": "test-trail-1", "has_custom_event_selectors": false, "tags": {}, "include_global_service_events": true, "s3_bucket_name": "cn-sh-cnn1-cloud-trail", "home_region": "cn-north-1", "is_logging": true, "is_multi_region_trail": false, "trail_arn": "arn:aws:cloudtrail:cn-north-1:640673851126:trail/test-trail-1"}, "changed": true, "exists": false}

***********************************
PARSED OUTPUT
{
    "changed": true,
    "exists": false,
    "invocation": {
        "module_args": {
            "aws_access_key": null,
            "aws_secret_key": null,
            "cloudwatch_logs_log_group_arn": null,
            "cloudwatch_logs_role_arn": null,
            "ec2_url": null,
            "enable_log_file_validation": null,
            "enable_logging": true,
            "include_global_events": true,
            "is_multi_region_trail": false,
            "kms_key_id": null,
            "name": "test-trail-1",
            "profile": "<redacted>",
            "region": "cn-north-1",
            "s3_bucket_name": "cn-sh-cnn1-cloud-trail",
            "s3_key_prefix": null,
            "security_token": null,
            "sns_topic_name": null,
            "state": "present",
            "tags": {},
            "validate_certs": true
        }
    },
    "trail": {
        "has_custom_event_selectors": false,
        "home_region": "cn-north-1",
        "include_global_service_events": true,
        "is_logging": true,
        "is_multi_region_trail": false,
        "name": "test-trail-1",
        "s3_bucket_name": "cn-sh-cnn1-cloud-trail",
        "tags": {},
        "trail_arn": "arn:aws:cloudtrail:cn-north-1:640673851126:trail/test-trail-1"
    }
}
```

note: "trail_arn": "arn:aws:cloudtrail:cn-north-1:640673851126:trail/test-trail-1" isn't a possible ARN, the partition name would be `aws-cn` for a real trail.

kwerey replied:
Ha, no problem. Thanks for picking up bug reports that affect a tiiiiiiny percentage of Cloudtrail users :)

tedder commented:
`shipit`

s-hertel commented:
bot_status

kwerey commented:
ready_for_review

kwerey commented:
@s-hertel thanks for poking the bot about this PR. I'd figured since a maintainer marked it as shippable it didn't need any further github workflow wrangling - is anything else needed to move it forward?

kwerey commented:
Taking a second look - @tedder , my guess is the magical ansibot will wait on you to change your review status from 'requested changes' to 'approved' before merging this in.

s-hertel commented:
@kwerey Nothing needed, this looks great. Thanks for fixing this. I just wanted to poke the bot to make sure it was safe to merge. :)

tedder commented:
Thanks s- and k.
